Authenticate embedded Zupass using localStorage via StorageAccessAPI

robknight opened the pull request:
Closes https://linear.app/0xparc-pcd/issue/0XP-1111/log-in-via-cookie
Closes https://linear.app/0xparc-pcd/issue/0XP-1139/use-requeststorageaccess-to-access-local-storage

In most browsers, local storage is "partitioned" by top-level origin. This means that if I visit zupass.org, I can create cookies, local storage entries, and so on, and these will be available to other tabs or windows pointing to zupass.org (in the same browser).

However, iframes do not work like this. They will not, by default, share local storage with other instances. To date, we simply allow the user to log in via the iframe, which downloads their storage. However, requiring the user to enter their username and password adds some friction.

~The only universally-supported way around this is to call `requestStorageAccess`, which grants access to first-party _cookies_ in a third-party (iframe) context. In this PR, Zupass sets a cookie containing the user's encryption key when they log in. When Zupass is loaded in an iframe, storage access is requested and, if approved, is used to read the encryption key from the cookie and log the user in.~

Using cookies to transfer information to iframes is not sufficiently secure; see comments below.

However, we can use a Chrome-specific API which allows iframes to access first-party local storage, with user permission. This allows us to retrieve the information necessary to boostrap a user session. This will only work in Chrome and derivatives like Arc, but since this is around 66% of browsers by market share, it seems worth doing.

This reduces the interaction required when using Zupass in an iframe from entering an email and password and clicking a button to simply clicking a button and clicking a further confirmation to grant storage access.

In cases where we detect that the user is running Chrome, instead of showing a request for email and password, we show this:

<img width="517" alt="image" src="https://github.com/user-attachments/assets/900768fd-d810-4664-ad13-a6d50975d5bd">

Clicking the button will request storage access:
<img width="941" alt="image" src="https://github.com/user-attachments/assets/c47d8aba-1fb3-4655-b5d5-e227c5c5a1b2">


If the user approves, and the local storage contains a valid encryption key, then we hash the encryption key and request the encrypted storage from the server. If this decrypts successfully then we use the storage to log the user in.

On subsequent visits, the embedded Zupass will remember the user and will not require clicking a button or logging in.

artwyman commented:
I may be showing my ignorance of web stuff here, but isn't the point of cookies that they are sent to the server on HTTP requests?  That's usually how server-side auth works on many cloud services.  But the user's encryption key is explicitly never supposed to be available to the server.  So even if our server code happens to never be looking for the cookie, it contains data which the client shouldn't allow the server to see.

I tried looking up the attributes to see if they change this.  As best I could Google, `SameSite=None` means the cookie is always sent (there's no option for never).  `Secure` only means the cookie is sent only over HTTPS, which we're always using anyway.

Am I missing something fundamental here?

robknight commented:
> I may be showing my ignorance of web stuff here, but isn't the point of cookies that they are sent to the server on HTTP requests? That's usually how server-side auth works on many cloud services. But the user's encryption key is explicitly never supposed to be available to the server. So even if our server code happens to never be looking for the cookie, it contains data which the client shouldn't allow the server to see.
> 
> I tried looking up the attributes to see if they change this. As best I could Google, `SameSite=None` means the cookie is always sent (there's no option for never). `Secure` only means the cookie is sent only over HTTPS, which we're always using anyway.
> 
> Am I missing something fundamental here?

Hmm. If we really do want to ensure that the key can't be sent to the server then this would seem to be a non-starter. I had assumed that we must be sending it, due to the fact that it's passed to `requestDownloadAndDecryptUpdatedStorage`, and `request*` functions are normally thin wrappers over "send these values to the server and return the result". In fact it does some hashing and decryption logic inside the API wrapper, which is surprising to me.

robknight commented:
Ah, I see the problem: for new users, we really do send the encryption key to the server, and store it in the database. However, this is effectively a temporary measure, until the user sets a password, after which point the server-side encryption key is deleted, and the user-generated key (derived from their password) is never sent to the server.

Hmm. This would seem to block cookies as a channel for sharing authentication state. Chrome does support sharing of first-party local storage with third-party contexts, so we can still make something work in Chrome, but not in other browsers (until the extended API in Chrome becomes a standard, if it does).

artwyman commented:
> Ah, I see the problem: for new users, we really do send the encryption key to the server, and store it in the database. However, this is effectively a temporary measure, until the user sets a password, after which point the server-side encryption key is deleted, and the user-generated key (derived from their password) is never sent to the server.

Right, that's a less-secure alternative for users who don't want to set a password.  For users with a password, E2EE implies that the server never sees the encryption key.  There might be something clever we could pass through a cookie to get the user authenticated.  We could make a special-case for users without passwords if that seemed worthwhile.  For users with a password, it seems like they either need to type the password into an iframe, or maybe do a popup-based flow which would send back the encryption key in client-only URL param.  The latter could work well if we're willing to have Zapps go through something akin to an OAuth flow.

From your description, I now separately wonder whether it's actually possible to create a new account where the encryption key is never seen by the server (as opposed to seen then deleted), but that's a separate discussion.

artwyman commented on apps/passport-client/components/screens/LoginScreens/LoginScreen.tsx:
```diff
@@ -126,6 +130,78 @@ export function LoginScreen(): JSX.Element {
+          return tryToLogin(encryptionKey);
+        }
+      })
+      .catch(() => {})
```
Is this line what's expected to handle the case of non-Chrome browsers?  What does the user see in this case?  Normal login screen in the iframe?

robknight replied:
If we're in a non-Chrome browser then this request will not have happened - all non-Chrome browsers return on line 168.

This catches the exception thrown if the user rejects the storage request, or has some browser setting or extension which causes it to be rejected. By calling `setTryStorageAccess(false)` we turn off the display of the storage access request and return the user to the regular login flow.

robknight commented:
I've added some new comments which hopefully make the flow clearer. I've also moved the `isChrome` check to earlier in the process, because we can avoid any checks for storage access if we know that the browser is not Chrome.

artwyman commented on apps/passport-client/components/screens/LoginScreens/LoginScreen.tsx:
```diff
@@ -126,6 +130,93 @@ export function LoginScreen(): JSX.Element {
+        appConfig.zupassServer,
+        encryptionKey
+      );
+      if (storageRequest.success) {
```
What happens if this fails?  There's no signaling of an error to the user, or returning of a value here, which is surprising.  Seems like this needs similar error handling, and maybe a loading spinner during download, similar to other pages which do this download and login.

robknight replied:
I've added an error message in the case where this fails.

artwyman commented on apps/passport-client/components/screens/LoginScreens/LoginScreen.tsx:
```diff
@@ -126,6 +130,93 @@ export function LoginScreen(): JSX.Element {
+        } else {
+          // We have access, try logging in
+          // Show a spinner:
+          setTryStorageLogin(true);
```
Also setTryStorageAccess(false), just to be sure your state-machine transitions are clear?
Or maybe these two booleans really should be an enum representing a state-machine directly.

robknight replied:
I've refactored the state machine into an enum with explicit states.

artwyman commented on apps/passport-client/components/screens/LoginScreens/LoginScreen.tsx:
```diff
@@ -126,6 +130,93 @@ export function LoginScreen(): JSX.Element {
+      const parser = new UAParser();
+      const browserName = parser.getBrowser().name;
+      const browserVersion = parser.getBrowser().version;
+      const isChrome125OrAbove =
```
This should be factored out into a helper function (likely in a different file) like checkStorageAccesAPI.

Do you know how this behaves for things which are Chrome-based but not Chrome?  E.g. Microsoft Edge?  Or embedded web apps based on Chromium such as Electron?  If those other situations have the storage access API, and we can properly detect them here, it broadens the set of users/cases we can handle.  The docs I can find suggest that some non-Chrome browsers like Opera already support this: https://developer.mozilla.org/en-US/docs/Web/API/Document/requestStorageAccess

It would be good if we could directly detect the presence of the feature rather than trying to figure out the browser version, but I don't see an obvious way to do that.  We can't just whether `document.hasStorageAccess` and `document.requestStorageAccess` are defined, because they always are, and what we need to know is whether the types.localStorage field is supported.

robknight replied:
I've tested UAParser with other Chrome-based browsers and it generally returns different names for browsers which are Chrome-based not actually Chrome (e.g. Edge, Brave). Interestingly Arc is identified as Chrome, but this seems to be because it really is just Chrome with a new UI, and has an identical feature-set to Chrome given equivalent version numbers. The `requestStorageAccess` function works identically in Arc as it does in Chrome.

artwyman replied:
Oh, and I see in the docs that Edge doesn't actually support local storage access, so I guess we don't want to know if it's "Chrome based".
A potential solution here is "just ask and see what we get", but that has the downside that the user sees a request.
That approach might work out if we were requesting cookie access anyway for some useful purpose (like skipping email verification as we discussed elsewhere), in which case doing the local storage access request could be an optional part of the same user interaction.  But that's an option for later, not this PR.

artwyman commented on apps/passport-client/components/screens/LoginScreens/LoginScreen.tsx:
```diff
@@ -126,6 +130,93 @@ export function LoginScreen(): JSX.Element {
+        await document.requestStorageAccess({ localStorage: true });
+
+      // Access granted, try reading the local storage
+      const encryptionKey = handle.localStorage.getItem("encryption_key");
```
It would be nice if this could just go through a normal load using the existing local storage rather than doing an extra download + login and creating a separate "device" for this iframe.  The separate local storage is going to create more opportunities for sync conflicts and stale data.

That being said, it looks like doing that might be complicated, since there are multiple things to request access to beyond local storage, and I'm not sure which of this will cause something to unexpectedly break in future (e.g. broadcast channels for talking to the service worker, or shared workers if we use them in future).  This approach seems sound for now.

robknight replied:
I think we could revisit this if it seems like it's a necessary optimization. It might be useful to cut down on the amount of sync traffic - I'll add a Linear issue for this.

artwyman replied:
It would avoid a download, but the sync traffic after this point would be the same.
It would importantly also reduce the number of copies of Zupass state stored on the user's device, which could be a size concern.
If there's a future version of sync which knows how to load and merge from LocalStorage then there would be a bigger difference in behavior, but that's theoretical.
None of these are important concerns for now.

artwyman commented on apps/passport-client/components/screens/LoginScreens/LoginScreen.tsx:
```diff
@@ -166,7 +257,23 @@ export function LoginScreen(): JSX.Element {
+      {tryStorageAccess && (
+        <TextCenter>
+          <Spacer h={24} />
+          Do you want to connect to Zupass?
```
This text should really be more specific about the ZApp in which the user is connecting.
At minimum it seems like it should say something about the URL of window.parent so the user sees what they're connecting to Zupass.
Out of the scope of this PR, this would be a place to plug into more information (like a known app or developer name) from a registry of approved Zapps if we had one.  That would come after any security check to verify that the Zapp should be allowed to connect at all.  I think we do need at least some form of that (like a whitelist of hostnames) before ZApps go to prod for all users, so that we ensure someone can't create a phishing site which embeds Zupass in an iframe and steals all a user's PODs.

robknight replied:
I've added specific text explaining which Zapp is connecting.

artwyman reviewed (changes requested):
This largely makes sense to me.  Inline comments are a few concrete suggestions, but also many points for dicussion of possibilities.

artwyman commented on apps/passport-client/components/screens/LoginScreens/LoginScreen.tsx:
```diff
@@ -154,6 +165,12 @@ export function LoginScreen(): JSX.Element {
           storage: storageRequest.value,
           encryptionKey
         });
+      } else {
```
I think it would be cleaner/clearer if this function returned a boolean and the else clause happened in the caller, but I think the current logic is correct.

artwyman commented on apps/passport-client/components/screens/LoginScreens/LoginScreen.tsx:
```diff
@@ -164,23 +181,28 @@ export function LoginScreen(): JSX.Element {
     } finally {
-      setTryStorageAccess(false);
-      setTryStorageLogin(false);
+      setStorageAccessStatus(StorageAccessStatus.Denied);
```
This is going to override values set in various codepaths above (including Success or NoLocalStorage).
I think this line belongs in the catch block, not the finally block.
Or maybe you want it to be `if (state===Requesting) { state = Denied; }` as a fallback, but at the moment I don't think there's a non-exception case which would trigger that.

robknight replied:
Yes, this belongs in the catch block - I've moved it. Thanks!

artwyman reviewed (changes requested):
One logic bug I'm concerned about (the finally block) but otherwise looks good.
